src/calibrationlib/csv_writer.cpp

```diff
@@ -0,0 +1,114 @@
+#include "csv_writer.hpp"
+
+#include <map>
+#include <string>
+#include <utility>
+
+#include <fstream>
+
+#include <boost/algorithm/string.hpp>
+
+namespace opossum {
+
+CSVWriter::CSVWriter(
+    const std::string file_path,
+    const std::vector<std::string>
+        headers,  // headers for csv file, if the file for file_path already exists, it must have equal headers
+    bool override_file)
+    : _headers(headers), _file_path(std::move(file_path)) {
+  // Get path from string
+  std::filesystem::path path_to_csv(_file_path);
+
+  // Extract directory path from file
+  const auto directory_path = path_to_csv.parent_path();
+
+  // Check if directory already exists, if not create the directory
+  if (!std::filesystem::exists(directory_path)) {
+    std::filesystem::create_directories(directory_path);
+  }
+
+  // Check if directory_path is actually a directory.
+  // If we do not create the directory, we cannot be sure if the provided path is valid
+  DebugAssert(std::filesystem::is_directory(directory_path), directory_path.string() + " is not a directory.");
+
+  // If we want to override the existing file, create a new one and insert headers.
+  if (override_file) {
+    _create_file_with_headers();
+  }
+
+  // Check existing file header is equal to new header in memory (important if replace_file = false)
+  DebugAssert(([&]() -> bool {
+                std::ifstream check_file;
+                check_file.open(_file_path);
+                std::string first_line;
+
+                auto correct_header = false;
+
+                getline(check_file, first_line);
+                {
+                  std::vector<std::string> to_test_headers;
+                  boost::split(to_test_headers, first_line, [&](char c) { return c == _delimiter; });
+
+                  if (to_test_headers == _headers) {
+                    correct_header = true;
+                  }
+                }
+                check_file.close();
+                return correct_header;
+              })(),
+              "\nCSV Writer ERROR: Provided headers for file '" + _file_path +
+                  "' does not match with existing headers. \n" + "Please validate headers for this CSVFile.");
```

Comment:
"CSV file"

Looks like 80% of that method is a DebugAssert. Sure that's a good idea?

Reply:
The DebugAssert checks whether the headers of an existing file equals the headers provided in the constructor of the class. 

To validate that, I have to do the following steps

- Open the file
- Read the first line
- Comparing headers. This can be done by:
  - Split the line into values and comparing those with the existing vector of headers
  - OR: comparing the entire string of the first line with a computed header line
- Close File

The comparing part does need some additional code in both cases. If you have any idea to shorten this DebugAssert, please let me know.

Reply:
1. What is the scenario where we already have a file with the same header and do not want to overwrite it? Do we want to append to this file?
2. Why should that code only be relevant for Debug builds?

Reply:
And what are "provided headers" for a file?

Reply:
1. Yep, this is relevant if we want to append data points to the file. (I have to admit that this is not the case at the moment, but I thought this might be useful in the future, which is why I added it)
2. Good question! I struggled with this decision. The main thought behind it was that the calibration is tested in debug mode beforehand. Asserts, which are not important should be skipped in the release build to improve performance. I agree, that this DebugAssert might be useful in release builds. What do you think?

Provided headers --> `headers` parameter of the CSVWriter constructor

```cpp
 CSVWriter(const std::string file_path, const std::vector<std::string> headers, const bool override_file = true);

```

PS: I wrote DebugAssert's for this class mainly because I ran into a lot of corrupted CSV files because values not being added mainly due to if statements without a default value.

Reply:
In order to ease the code and ensure that everything that should be supported is actually implemented and tested, I'd vote for removing all capabilities to append to CSVs.
It's all or nothing.

Reply:
Appending is tested and implemented, just not used in other classes. Nevertheless, I removed the feature to append to existing files in [
315d2e0](https://github.com/hyrise/hyrise/pull/2009/commits/315d2e08b2aaf30628adb3d684c4bf09ad198118)